src/pvcam.cpp

```diff
@@ -0,0 +1,1200 @@
+    frame_count_ = 0;
+
+    // start acquisition
+    PVCAM(
+      pl_exp_start_cont(hcam_, (void*)pl_buffer_, (uint32_t)pl_buffer_size_),
+      *pvcam_api_mutex_);
+
+    int16_t status;
+    uint32_t byte_cnt, buffer_cnt;
+    PVCAM(pl_exp_check_cont_status(hcam_, &status, &byte_cnt, &buffer_cnt),
+          *pvcam_api_mutex_);
+
+    started_ = true;
+}
+
+void
+PVCamCamera::stop()
+{
+    if (!started_) {
+        return;
+    }
+
+    LOG("Stopping");
+    PVCAM(pl_exp_stop_cont(hcam_, CCS_HALT), *pvcam_api_mutex_);
+    started_ = false;
+
+    delete[] pl_buffer_;
+    pl_buffer_ = nullptr;
+    pl_buffer_size_ = 0;
+}
+
+void
+PVCamCamera::execute_trigger() const
+{
+    // In current implementation the input flags should be set to 0. On output,
+    // the flags will contain one of the values defined in PL_SW_TRIG_STATUSES
+    // enumeration.
+    uint32_t flags = 0;
+    PVCAM(pl_exp_trigger(hcam_, &flags, 0), *pvcam_api_mutex_);
+
+    if (flags != PL_SW_TRIG_STATUS_TRIGGERED) {
+        LOGE("The camera was unable to accept the trigger due to an ongoing "
+             "exposure.");
+    }
+}
+
+void
+PVCamCamera::get_frame(void* im, size_t* nbytes, struct ImageInfo* info)
+{
+    CHECK(im);
+    CHECK(nbytes);
+    CHECK(info);
+
+    int16_t status;
+    uint32_t byte_cnt, buffer_cnt;
+    PVCAM(pl_exp_check_cont_status(hcam_, &status, &byte_cnt, &buffer_cnt),
+          *pvcam_api_mutex_);
+
+    std::unique_lock<std::mutex> frame_lock(*frame_mutex_);
+
+    // callback will increment frames_available
+    callback_context_.cv.wait_for(
+      frame_lock,
+      std::chrono::microseconds(exposure_time_ * exposure_time_to_us_),
+      [&] { return callback_context_.frames_available > 0; });
+
+    if (0 == callback_context_.frames_available) {
+        *nbytes = 0;
+        return;
+    }
+    --callback_context_.frames_available;
+
+    uint32_t width = last_known_settings_.shape.x,
+             height = last_known_settings_.shape.y;
+    SampleType pixel_type = last_known_settings_.pixel_type;
+
+    memcpy(
+      im, *callback_context_.frame, width * height * bytes_of_type(pixel_type));
+
+    *info = {
+            .shape = {
+              .dims = {
+                .channels = 1,
+                .width = width,
+                .height = height,
+                .planes = 1,
+              },
+              .strides = {
+                .channels = 1,
+                .width = 1,
+                .height = width,
+                .planes = width * height,
+              },
+              .type = pixel_type,
+            },
+            .hardware_timestamp = (uint64_t)callback_context_.info.TimeStamp,
+            .hardware_frame_id = (uint64_t)callback_context_.info.FrameNr,
+        };
+
+    PVCAM(pl_exp_unlock_oldest_frame(hcam_), *pvcam_api_mutex_);
+
+    ++frame_count_;
+
+    rgn_type roi;
+    PVCAM(pl_get_param(hcam_, PARAM_ROI, ATTR_CURRENT, &roi),
+          *pvcam_api_mutex_);
```

Comment:
Why get the ROI here? Does this have some kind of positive side-effect? Or maybe it's a remnant from testing?

Reply:
Yeah, this looks like a holdover from testing. Thanks for catching.